### DISCUSSION TITLE
Monthly Pages

### DISCUSSION THREAD

**Discussion by stdword** [stdword/logseq13-full-house-plugin#44] [Showroom]
###### inspired by [this](https://www.youtube.com/watch?v=Q7QVVIRnKoI) Obsidian video and @inzagi1's [question](https://github.com/stdword/logseq13-full-house-plugin/discussions/42)

<video width="800px" src="https://github.com/user-attachments/assets/c7c61443-4bbd-40bb-8470-4810e497912a"></video>

## Steps

0) Setup «Weekly Pages» with [this](https://github.com/stdword/logseq13-full-house-plugin/discussions/43) manual.
1) Install [`Side Block`](https://github.com/YU000jp/logseq-plugin-side-block) plugin to create *side views*.
2) Let's prepare the queries. Create anywhere in Logseq (ex: in «Queries» page) the following queries. I'll use 2 queries for \#podcast and \#diary records and the one for . But you can setup your own ones:
   ````org-mode
   reference:: #podcast
   #+BEGIN_QUERY
   …
   #+END_QUERY
   ````
   ````markdown
   reference:: #diary
   #+BEGIN_QUERY
   …
   #+END_QUERY
   ````
   Fill the body of queries with the same one code. Insert it between `#+BEGIN_QUERY` and `#+END_QUERY` with replacing the «…» (use copy button in the top right corner):
   ````clojure
   {
    :inputs [:current-page :current-block]
    :query [:find (pull ?b [*])
    :in $ ?month ?self %
    :where
      [?page :block/name ?month]
      (get-month ?page ?monthStart ?monthEnd)
      (between ?b ?monthStart ?monthEnd)
      
      [?self :block/page ?self-page]
      (tagged-with-property ?b ?self :reference)
    ]
    :rules [
      [(get-month ?block ?start ?end)
        [?block :block/properties ?props]
        [(get ?props :month-start) ?start]
        [(get ?props :month-end) ?end]
      ],
      [(tagged-with-property ?block ?props-block ?prop)
        [?props-block :block/properties-text-values ?props]
        [(get ?props ?prop) ?tag]
        [(subs ?tag 1) ?tag-name]
        [?block :block/refs ?ref]
        [?ref :block/name ?name]
        [(= ?name ?tag-name)]
      ]
    ]
   }
   ````
   **Restriction**: currently there is no way in Logseq to avoid copying of query body. Please vote for this feature [here](https://discuss.logseq.com/t/make-advanced-queries-reusable) and like [this](https://discuss.logseq.com/t/make-advanced-queries-reusable/15281/13?u=stdword) comment.
2) Create the query for getting #weekly records from Weekly Pages (use copy button in the top right corner):
   ````clojure
   #+BEGIN_QUERY
   {
    :collapsed? false
    :inputs [:current-block :current-page]
    :query [
      :find (pull ?b [*])
      :in $ ?self ?name %
      :where
        [?b :block/page ?p]
        (referenced-by-property ?p :month ?name)
        (tagged-with-property ?b ?self :reference)
    ]
    :rules [
        [(tagged-with-property ?block ?props-block ?prop)
          [?props-block :block/properties-text-values ?props]
          [(get ?props ?prop) ?tag]
          [(subs ?tag 1) ?tag-name]
          [?block :block/refs ?ref]
          [?ref :block/name ?name]
          [(= ?name ?tag-name)]
        ],
        [(referenced-by-property ?entity ?prop ?ref)
          [?entity :block/properties ?props]
          [(get ?props ?prop) ?val]
          [(contains? ?val ?ref)]
        ]
    ]
   }
   #+END_QUERY
   reference:: #weekly
   ````

3) Create `🏛template` anywhere in Logseq (ex: in «Templates» page). (Use copy button in the top right corner):
   ````markdown
   - template:: monthly page
     template-version:: 2
     plugin-version:: 4.0.0
     link:: [Monthly Pages](https://github.com/stdword/logseq13-full-house-plugin/discussions/44)
     arg-day-format:: MMM, Do
     arg-week-format:: YYYY-[W]W
     collapsed:: true
     - ```javascript
       ``{
           var humanFormat = d => d.format(c.args['day-format'])
           var weekFormat = d => d.format(c.args['week-format'])
           var queryFormat = d => d.format('YYYYMMDD')
           
           var m = c.page.name.match(/(\d{4}).*?(\d{1,2})/)
           if (!m) {
               await logseq.UI.show_msg('This is not a monthly page')
               return
           }
           var [year, month] = [m[1], m[2]]
           
           var start = date.from(c.page.name).startOf('month')
           var end = start.endOf('month')
         
           // set page properties
           outn('year:: ' + ref(year))
           outn('type:: month')
           outn('month-start:: ' + queryFormat(start))
           outn('month-end:: ' + queryFormat(end))
         
           // get all weeks
           var weeks = []
           var weekStart = start.startOf('isoWeek')
           while (true) {
               if (weekStart.isAfter(start, 'month'))
                   break
               var weekEnd = weekStart.endOf('isoWeek')
               var week = Number(weekStart.format('W'))
           weeks.push([week, weekStart, weekEnd])
               weekStart = weekStart.add(1, 'week')
           }
         
           // prepare «by week»  overview block
           var getByWeekQuery = function (week, title, start, end) {
             return `
       #+BEGIN_QUERY
       {:title [:p [:font {:size 4} [:b "W${week}"]] "  " [:i "${title}"]]
        :collapsed? true
        :inputs [:current-block ${start} ${end}]
        :query [
          :find (pull ?b [*])
          :in $ ?self ?start ?end %
          :where
            (between ?b ?start ?end)
            (tagged-with-property ?b ?self :reference)
        ]
        :rules [
            [(tagged-with-property ?block ?props-block ?prop)
              [?props-block :block/properties-text-values ?props]
              [(get ?props ?prop) ?tag]
              [(subs ?tag 1) ?tag-name]
              [?block :block/refs ?ref]
              [?ref :block/name ?name]
              [(= ?name ?tag-name)]
            ]
        ]
       }
       #+END_QUERY
           `.trim()
           }
           
           var generateByWeekBlock = function () {
             for (const [week, start, end] of weeks) {
                 var title = `${humanFormat(start)} - ${humanFormat(end)}`
                 var weekQuery = getByWeekQuery(week, title, queryFormat(start), queryFormat(end))
                 outn(weekQuery.trim())
             }
           }
       }``
       ```
     - Day & Week journals
       heading:: true
       - {{renderer :days, *, -, ``year``, ``month``}} #.side
         - INSERT HERE embed ref to #weekly query
           reference:: #weekly
           query-table:: true
           query-sort-by:: page
           query-sort-desc:: false
     - Podcasts overview
       heading:: true
       - {{renderer :days, [[podcast]], -, ``year``, ``month``}} #.side
         - INSERT HERE embed ref to #podcast query
           reference:: #podcast
       - By weeks
         heading:: true
         collapsed:: true
         - `{ generateByWeekBlock() }`
           reference:: #podcast
     - Diary overview
       heading:: true
       - {{renderer :days, [[diary]], -, ``year``, ``month``}} #.side
         - INSERT HERE embed ref to #diary query
           reference:: #diary
       - By weeks
         heading:: true
         collapsed:: true
         - ``{ generateByWeekBlock() }``
           reference:: #diary
   ````
4) Copy reference to every created before query via right clicking on corresponding bullet and selecting «Copy block embed»:
   <img width="300px" src="https://github.com/stdword/logseq13-full-house-plugin/assets/1984175/f43b3673-5baf-4c90-a493-e10c98309a41"/>
   …and insert them replacing line with «INSERT HERE» text in the template (there are three lines). It should look like:
   ```
   …
   - {{embed ((666cd068-133c-4b30-9df4-1dd8a94179cb))}}
   …
   ``` 
5) Now, let's install and setup [`Default Template`](https://github.com/YU000jp/logseq-plugin-default-template) plugin to automatically insert `🏛️template`. Create another template:
   ```mardown
   - template:: .monthly page inserter
     template-including-parent:: false
     - {{renderer :template, monthly page, :delay-until-rendered}}
     - 
   ```
   And add `\d{4}-\d{2}::.monthly page inserter` text to `Default Template` plugin's setting `/advancedRegex`:
   <img width="400" src="https://github.com/user-attachments/assets/fea2922e-4244-4141-ab99-3aff4f1c762c">
6) Additionally you can setup [`Awesome Links`](https://github.com/yoyurec/logseq-awesome-links) plugin to have colored *references* to Monthly Pages (according to `type:: month`):
   - Goto plugin settings (`t s`) → *Awesome Links* and set the `inheritFromProp` setting to `type`:
     <img width="400" src="https://github.com/user-attachments/assets/0f854e8a-6d3b-4ad9-902b-10b37308504d">
   - Go to `[[month]]` page and add page property `color:: mediumseagreen` (choose any value you want via [this service](https://enes.in/sorted-colors)).

❤️ Enjoy the full automatic monthly page creation!

### Why are queries embedded instead of being written directly in the template?
Because if you want to change the query, you would only need to modify it in one place - and the changes will be immediately visible on all already created weekly pages. This is very handy. But if you don't need this - you can do without _embed_ and each time "enclose" the query inside the weekly page. Then, when changing the query in the template, only new pages will see its new look.

### Why is the property `reference::` with the value `#podcast` displayed in the query block?
1. Visually, it's clear what type of data we're looking at. Although you could use the query title for this, or simply write the heading "Podcasts Overview".
2. For an easy configuration, if you suddenly want to change `#podcast` to something else (for example, to `#youtube-video`).
3. To have a block with a query in _Linked References_, on the `podcast` page. This is handy if you want to rename the `podcast` page (e.g. to `🎙️`). The block with the query will be renamed **automatically**. If it is not in _Linked References_, you will have to a) remember that the block with the query on the pages requires manual modification b) carry out this manual modification.
- If you want to hide this property, you can do this via `custom.css`:
  ```css
  .block-properties > div:has(.page-property-key[data-ref="reference"]) {
      display: none !important;
  }
  ```
  At the same time, the property continues to be visible in edit mode, which is handy if you need to change its value.

## See also
- [Weekly Pages](https://github.com/stdword/logseq13-full-house-plugin/discussions/43) setup

## Additional info
- Logseq v0.10.9
- ☕️ [BuJo Theme](https://github.com/stdword/logseq-bujo-theme)
- [discuss.logseq.com](https://discuss.logseq.com/t/monthly-journal-pages/28106)

**Comment by YU000jp** [comment]
One idea has occurred to me. To make this a homepage, it might be useful to have a plugin that redirects to that month's page immediately when a temporary page is accessed!

**Reply by stdword** [reply]
Nice idea! 🔥

**Comment by MarcoBarna** [comment]
hi @stdword i'm trying to implement the monthly and yearly pages but the side calendar doesn't show and the generateByWeekBlock is not executed

**Reply by stdword** [reply]
Hello, @MarcoBarna! Thank you for a feedback 🙏
It was a bug and I've just fixed the code to copy.

You can repeat whole instruction steps or change the template according to this video:

https://github.com/user-attachments/assets/6e17318a-69ea-49f2-b232-816f05dbc85e

**Comment by prashantsengar** [comment]
Is it possible to use simple queries and use the variables? 

For example, I use the smartblocks plugin to create monthly page and my query looks like this:

```
{{query (and (between <%getinput: startDate%> <%getinput: endDate%> ) [[Gratitude]] )}}
```

**Reply by stdword** [reply]
Hello!
No, there is no such option like in SmartBlocks plugin.

But you could use:
1. [Arguments](https://stdword.github.io/logseq13-full-house-plugin/#/reference__args) for the template / view with the way to set [default ones](https://stdword.github.io/logseq13-full-house-plugin/#/reference__configuring?id=default-usage) to achieve the same
2. [date.nlp](https://stdword.github.io/logseq13-full-house-plugin/#/reference__tags?id=date-nlp) template tag for easier setting up the exact dates.
2. [Query language for pages](https://stdword.github.io/logseq13-full-house-plugin/#/reference__query_language) if you want to construct query for pages (blocks doesn't supported for now).
